### DISCUSSION TITLE
Should we split `network` into a separate crate

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1859] [Ideas]
The `network` module does not get a lot of attention from us. As part of the effort to stabalize the `rust-bitcoin` crate would it be useful to pull `network` out into a separate crate.

Observe that the `network` module depends on the rest of `rust-bitocin`, there is no dependency by the rest of `rust-bitcoin` on the `network` module (after #1854).

**Comment by Kixunil** [comment]
More modules than just `network` don't get a lot of our attention. And it's pretty natural that we work on the core types currently - they are most often needed in our projects.

We should focus on stabilizing core types first: block data, `Amount`, `Address`, and hashes. Which means they will require separate crates.

**Reply by apoelstra** [reply]
Plausibly we could work on defining and stabilizing `primitives`, and then see what's even left over in the main `bitcoin` crate.

**Reply by Kixunil** [reply]
That's definitely the initial step. But I'd like to parallelize it a bit, since I want to have "no strong reason to change the API for a year" condition before stabilization and if we did it one after the other it'd take insane amount of time.

**Comment by tcharding** [comment]
I could have explained my motivation for raising this discussion and #1860 better. Given how long it takes us to make new "acceptable" crates (*cough* hex) I thought another way to attack the stabilisation task is to tease out things that don't need doing immediately (as opposed to creating things that are "acceptable". So `network` (after rename to `p2p`) could be put in a separate crate without changes and released as v0.1.0 this would:
- allow us to focus on `rust-bitcoin` crate
- signal to others that this crate is not as polished
- possibly allow others to work on the new crate and we could review/merge with less care/bikeshedding

Also it allows more work in parallel. Especially if review/ack requirements were lower. (Perhaps.)

These are all only vague feelings, totally open to ideas.

**Reply by Kixunil** [reply]
Yeah, this is pretty much what I was thinking about.

**Comment by Kixunil** [comment]
So based on @stevenroose answer, I propose to name the crate `bitcoin-p2p-types`. `bitcoin-wire` is also kinda good but I like `p2p-types` more because it's immediately obvious what the crate does.

**Reply by tcharding** [reply]
ref: https://github.com/stevenroose/rust-bitcoin-p2p/issues/3#issuecomment-1567660980

**Comment by tcharding** [comment]
+1 for `bitcoin-p2p-types`

Paraphrasing @stevenroose here to tie the two threads together. Steven suggests leaving the module in `rust-bitcoin`, note he is in camp "stop refactoring and leave shit as it is". I'm in camp "refactor everything and break it up so we can commit to a stable API". I think @Kixunil is in camp "break stuff up if its _really_ better", is that right?

I'll try to put a few hours into the `network` module with the all this in mind and see if we can find a compromise that works for everyone.

**Reply by Kixunil** [reply]
Yeah, pretty much.

**Reply by apoelstra** [reply]
I am also in camp "let's drive toward 1.0" and I think that will involve breaking a lot of shit whether I'm happy about it or not.

**Reply by stevenroose** [reply]
Let me know when we organize a water balloon fight between the camps to see who will win. I'm happily flying to Australia for that.

**Comment by stevenroose** [comment]
As for structuring this module/crate, I would do away with all the `message::` submodules and just expose all messages under `message::`. I mean we could have the submodules for organization sake, but I wouldn't expose them.

**Comment by tcharding** [comment]
Ah cool, related to this is https://github.com/rust-bitcoin/rust-bitcoin/pull/525 (the message parts). I was going to attempt that after #1854 was resolved. I'll add your comment to my todo list to do at the same time. Thanks
